File: ProblemSolving/450DSA/Python/src/dynamicprogramming/BBTCounter.py
"""
@author Anirudh Sharma

Given a height h, count the maximum number of balanced binary trees possible with height h.
Print the result modulo 10^9 + 7.
"""


def countBT(h):
    MODULUS = 1000000007
    # Lookup table to store intermediate results
    lookup = [0] * (h + 1)
    # Base initialization
    lookup[0] = 1
    lookup[1] = 1
    # Populate the remaining table
    for i in range(2, h + 1):
        lookup[i] = (lookup[i - 1] * ((2 * lookup[i - 2]) %
                                      MODULUS + lookup[i - 1]) % MODULUS) % MODULUS
    return lookup[h]


if __name__ == "__main__":
    print(countBT(2))
    print(countBT(3))

File: ProblemSolving/450DSA/JavaScript/src/dynamicprogramming/BBTCounter.js
/**
 * @author Anirudh Sharma
 * 
 * Given a height h, count the maximum number of balanced binary trees possible with height h.
 * Print the result modulo 10^9 + 7.
 */
const countBT = (h) => {
    const MODULUS = 1000000007;
    // Lookup table to store intermediate results
    const lookup = new Array(h + 1).fill(0);
    // Base initialization
    lookup[0] = 1;
    lookup[1] = 1;
    // Populate the remaining table
    for (let i = 2; i <= h; i++) {
        lookup[i] = (lookup[i - 1] * ((2 * lookup[i - 2]) % MODULUS + lookup[i - 1]) % MODULUS) % MODULUS;
    }
    return lookup[h];
};

const main = () => {
    console.log(countBT(2));
    console.log(countBT(3));
};

main();